### PR TITLE
Decoupling Stratified Split and Split

### PR DESCRIPTION
**Pull request by Abilityguy**
This PR tries to solve an issue encountered in PR #2746. (Refer this [comment](https://github.com/mlpack/mlpack/pull/2746#issuecomment-773230039) for more details.)

After a chat with @RishabhGarg108 on IRC, we decided it would be best to open a PR to look at the best way to solve this problem.

Currently, the ```StratifiedSplit``` method is linked to the ```Split``` method through the ```stratifyData``` argument. This PR separates the code section of ```StratifiedSplit``` into a separate file and introduces fixes in ```methods/preprocess/preprocess_split_main.cpp``` to accommodate this change. 

All related tests in ```tests/split_data_test.cpp``` and ```tests/main_tests/preprocess_split_test.cpp``` are passing on my local system.

A slight issue that's bugging me is the code repetition in the ```if (stratifyData)...else...``` section in ```preprocess_split_main.cpp```. If anyone has suggestions on how to implement this part better, that would be great.

### CODE REVIEW

**Review by RishabhGarg108 — APPROVED** [review]
Looks great to me! Thanks for the fix 👍

File: src/mlpack/core/data/split_data.hpp [hunk 1]
```diff
@@ -52,56 +54,27 @@ namespace data {
...
-   * Train set (4 0s, 9 1s)
-   * 0 0 0 0
-   * 1 1 1 1 1 1 1 1 1
-   *
-   * Test set (1 0s, 2 1s)
-   * 0
-   * 1 1
-   */
+  if (!arma::is_Row<LabelsType>::value)
+    throw std::runtime_error("data::Split(): when stratified sampling is done,"
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:69]:
```suggestion
    throw std::runtime_error("data::Split(): when stratified sampling is done, "
```

Missing space.

File: src/mlpack/core/data/split_data.hpp [hunk 2]
```diff
@@ -52,56 +54,27 @@ namespace data {
...
+        "labels must have type `arma::Row<>`!");
   size_t trainIdx = 0;
   size_t testIdx = 0;
   size_t trainSize = 0;
   size_t testSize = 0;
   arma::uvec labelCounts;
   arma::uvec testLabelCounts;
-  U maxLabel = inputLabel.max();
+  auto maxLabel = inputLabel.max();
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:77]: Do you mind to use `typename LabelsType::elem_type` here, for more details see the armadillo section about "Can I use the C++11 auto keyword with Armadillo objects and/or expressions?".

File: src/mlpack/core/data/split_data.hpp [hunk 3]
```diff
@@ -132,7 +105,7 @@ void StratifiedSplit(const arma::Mat<T>& input,
 
   for (arma::uword i : order)
   {
-    U label = inputLabel[i];
+    auto label = inputLabel[i];
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:108]: See comment above.

File: src/mlpack/core/data/split_data.hpp [hunk 4]
```diff
@@ -114,7 +87,7 @@ void StratifiedSplit(const arma::Mat<T>& input,
     order = arma::shuffle(order);
   }
 
-  for (U label : inputLabel)
+  for (auto label : inputLabel)
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:90]: See comment above.

File: src/mlpack/core/data/split_data.hpp [hunk 5]
```diff
@@ -368,6 +341,240 @@ Split(const arma::Mat<T>& input,
...
+           const bool shuffleData = true)
+{
+  const size_t testSize = static_cast<size_t>(input.n_cols * testRatio);
+  const size_t trainSize = input.n_cols - testSize;
+
+  trainData.set_size(1, trainSize);
+  testData.set_size(1, testSize);
+
+  arma::uvec order = arma::linspace<arma::uvec>(0, input.n_cols - 1,
+      input.n_cols);
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:399]:
```suggestion
      input.n_cols);

```

Insert extra line to improve readability.

File: src/mlpack/core/data/split_data.hpp [hunk 6]
```diff
@@ -368,6 +341,240 @@ Split(const arma::Mat<T>& input,
...
+  if (shuffleData)
+    order = arma::shuffle(order);
+
+  if (trainSize > 0)
+  {
+    trainLabels.set_size(1, trainSize);
+
+    for (size_t i = 0; i < trainSize; i++)
```
Comment (zoq) [inline src/mlpack/core/data/split_data.hpp:407]: Please use pre increment here, see https://github.com/mlpack/mlpack/pull/2464 for more details.

File: src/mlpack/tests/split_data_test.cpp [hunk 7]
```diff
@@ -322,3 +322,20 @@ TEST_CASE("StratifiedSplitLargerDataResultTest", "[SplitDataTest]")
   mat concat = arma::join_rows(std::get<0>(value), std::get<1>(value));
   CheckMatEqual(input, concat);
 }
+
+/**
+ * Check that Split() with stratifyData true throws a runtime error if labels
+ * are not of type arma::Row<>.
+ */
+TEST_CASE("StratifiedSplitRunTimeErrorTest", "[SplitDataTest]")
```
Comment (zoq) [inline src/mlpack/tests/split_data_test.cpp:330]: Can you add a test for the `field` data type as well?

File: src/mlpack/core/data/split_data.hpp (hunk 2 above)
Comment (Abilityguy) [reply src/mlpack/core/data/split_data.hpp:77]: Thanks for pointing this out!.

**Comment by Abilityguy** [comment]
Thanks for the review @zoq!.

The code from this PR has been cherry picked in PR #2746. I have requested for changes according to your review comments there.

I am closing this in favour of #2746 since it's better to have all the changes in a single PR.

**Comment by zoq** [comment]
> Thanks for the review @zoq!.
> 
> The code from this PR has been cherry picked in PR #2746. I have requested for changes according to your review comments there.
> 
> I am closing this in favour of #2746 since it's better to have all the changes in a single PR.

I see, thanks.

**Pull request closed without merging**